libcontainer/nsenter/nsexec.c

```diff
@@ -488,38 +496,148 @@ void join_namespaces(char *nslist)
 		strncpy(ns->type, namespace, PATH_MAX - 1);
 		strncpy(ns->path, path, PATH_MAX - 1);
 		ns->path[PATH_MAX - 1] = '\0';
+
+		ns_to_join |= nstype(ns->type);
 	} while ((namespace = strtok_r(NULL, ",", &saveptr)) != NULL);
 
-	/*
-	 * The ordering in which we join namespaces is important. We should
-	 * always join the user namespace *first*. This is all guaranteed
-	 * from the container_linux.go side of this, so we're just going to
-	 * follow the order given to us.
-	 */
+	*ns_list = namespaces;
+	*ns_len = len;
+	return ns_to_join;
+}
 
-	for (i = 0; i < num; i++) {
-		struct namespace_t *ns = &namespaces[i];
-		int flag = nsflag(ns->type);
+/*
+ * Try to join all namespaces that are in the "allow" nsset, and return the
+ * set we were able to successfully join. If a permission error is returned
+ * from nsset(2), the namespace is skipped (non-permission errors are fatal).
+ */
+static nsset_t __join_namespaces(nsset_t allow, struct namespace_t *ns_list, size_t ns_len)
+{
+	nsset_t joined = 0;
+
+	for (size_t i = 0; i < ns_len; i++) {
+		struct namespace_t *ns = &ns_list[i];
+		int type = nstype(ns->type);
+		int err, saved_errno;
 
-		write_log(DEBUG, "setns(%#x) into %s namespace (with path %s)", flag, ns->type, ns->path);
-		if (setns(ns->fd, flag) < 0)
+		if (!(type & allow)) {
+			write_log(DEBUG, "skipping setns(%#x) into %s namespace (with path %s)",
```

Comment:
Maybe we should check is this type ns has been joined before, or else it will make user confused for this log.

Reply:
Too many logs about skipping setns.
```bash
@lifubang ➜ ~/bb $ sudo /workspaces/runc/runc --debug exec -d test sleep inf
DEBU[0000]libcontainer/dmz/cloned_binary_linux.go:202 libcontainer/dmz.IsCloned() F_GET_SEALS on /proc/self/exe failed: invalid argument 
DEBU[0000]libcontainer/dmz/cloned_binary_linux.go:227 libcontainer/dmz.CloneSelfExe() runc-dmz: using overlayfs for sealed /proc/self/exe 
DEBU[0000]libcontainer/container_linux.go:506 libcontainer.(*Container).newParentProcess() runc-dmz: using /proc/self/exe clone         
DEBU[0000] nsexec[8341]: => nsexec container setup      
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec[8341]: set process as non-dumpable    
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-0[8341]: ~> nsexec stage-0            
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-0[8341]: spawn stage-1                
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-0[8341]: -> stage-1 synchronisation loop 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-1[8342]: ~> nsexec stage-1            
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-1[8342]: setns(0x8000000) into ipc namespace (with path /proc/8162/ns/ipc): Success 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-1[8342]: setns(0x4000000) into uts namespace (with path /proc/8162/ns/uts): Success 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-1[8342]: setns(0x40000000) into net namespace (with path /proc/8162/ns/net): Success 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-1[8342]: setns(0x20000000) into pid namespace (with path /proc/8162/ns/pid): Success 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-1[8342]: setns(0x20000) into mnt namespace (with path /proc/8162/ns/mnt): Success 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-1[8342]: setns(0x2000000) into cgroup namespace (with path /proc/8162/ns/cgroup): Success 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-1[8342]: skipping setns(0x8000000) into ipc namespace (with path /proc/8162/ns/ipc) 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-1[8342]: skipping setns(0x4000000) into uts namespace (with path /proc/8162/ns/uts) 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-1[8342]: skipping setns(0x40000000) into net namespace (with path /proc/8162/ns/net) 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-1[8342]: skipping setns(0x20000000) into pid namespace (with path /proc/8162/ns/pid) 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-1[8342]: skipping setns(0x20000) into mnt namespace (with path /proc/8162/ns/mnt) 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-1[8342]: skipping setns(0x2000000) into cgroup namespace (with path /proc/8162/ns/cgroup) 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-1[8342]: skipping setns(0x8000000) into ipc namespace (with path /proc/8162/ns/ipc) 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-1[8342]: skipping setns(0x4000000) into uts namespace (with path /proc/8162/ns/uts) 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-1[8342]: skipping setns(0x40000000) into net namespace (with path /proc/8162/ns/net) 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-1[8342]: skipping setns(0x20000000) into pid namespace (with path /proc/8162/ns/pid) 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-1[8342]: skipping setns(0x20000) into mnt namespace (with path /proc/8162/ns/mnt) 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-1[8342]: skipping setns(0x2000000) into cgroup namespace (with path /proc/8162/ns/cgroup) 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-1[8342]: unshare remaining namespaces 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-1[8342]: spawn stage-2                
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-1[8342]: request stage-0 to forward stage-2 pid (8343) 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-0[8341]: stage-1 requested pid to be forwarded 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-0[8341]: forward stage-1 (8342) and stage-2 (8343) pids to runc 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-1[8342]: signal completion to stage-0 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-0[8341]: stage-1 complete             
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-0[8341]: <- stage-1 synchronisation loop 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-0[8341]: -> stage-2 synchronisation loop 
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-0[8341]: signalling stage-2 to run    
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-1[8342]: <~ nsexec stage-1            
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-2[6]: ~> nsexec stage-2               
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() nsexec-2[6]: signal completion to stage-0    
DEBU[0000] nsexec-0[8341]: stage-2 complete             
DEBU[0000] nsexec-0[8341]: <- stage-2 synchronisation loop 
DEBU[0000] nsexec-0[8341]: <~ nsexec stage-0            
DEBU[0000] nsexec-2[6]: <= nsexec container setup       
DEBU[0000] nsexec-2[6]: booting up go runtime ...       
DEBU[0000] child process in init()                      
DEBU[0000] reading sync                                 
DEBU[0000] read sync type:procReady                     
DEBU[0000] writing sync type:procRun                    
DEBU[0000] reading sync                                 
DEBU[0000] sync pipe closed                             
DEBU[0000]libcontainer/logs/logs.go:55 libcontainer/logs.processEntry() writing sync type:procReady                  
DEBU[0000] reading sync                                 
DEBU[0000] read sync type:procRun                       
DEBU[0000] setns_init: about to exec
```

Reply:
We can probably skip the "skipping" messages, but we should keep the error and success lines. These are all debug logs, so there is basically no overhead to having them without debug logging enabled (we don't forward them in `log.c` if the log level isn't debug).